flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ProcessingTimeServiceImpl.java

```diff
@@ -39,11 +78,282 @@ public long getCurrentProcessingTime() {
+
+		private static final int STATUS_PENDING = 0;
+		private static final int STATUS_EXECUTING = 1;
+		private static final int STATUS_CANCELLED = 2;
+		private static final int STATUS_FAILED = 3;
+		private static final int STATUS_SUCCEEDED = 4;
+
+		// ------------------------------------------------------------------------
+
+		private final CompletableFuture<ScheduledFuture<?>> wrappedFuture;
+
+		private final boolean isPeriodic;
+
+		private final CompletableFuture<TimerScheduledFuture<?>> onDoneActionFuture;
+
+		/**
+		 * This is used to control execution behavior to ensure that the on-done action is executed
+		 * after {@link ProcessingTimeCallback#onProcessingTime(long)}.
+		 */
+		private final AtomicInteger status;
+
+		TimerScheduledFuture(boolean isPeriodic, Consumer<TimerScheduledFuture<?>> onDoneAction) {
+			this.isPeriodic = isPeriodic;
+
+			this.wrappedFuture = new CompletableFuture<>();
+			this.onDoneActionFuture = new CompletableFuture<>();
+			this.status = new AtomicInteger(STATUS_PENDING);
+
+			// if "onDoneAction" is executed in the timer thread, the execution of "onDoneAction" may
+			// be interrupted when canceling the timer, which will cause some state uncertainty, so we
+			// execute it asynchronously
+			onDoneActionFuture.thenAcceptAsync(checkNotNull(onDoneAction));
+		}
+
+		void bind(ScheduledFuture<?> future) {
+			checkState(!this.wrappedFuture.isDone());
+			this.wrappedFuture.complete(future);
+		}
+
+		ProcessingTimeCallback getCallback(ProcessingTimeCallback callback) {
+			return isPeriodic ? wrapPeriodicCallback(callback) : wrapOneShotCallback(callback);
+		}
+
+		private ProcessingTimeCallback wrapOneShotCallback(ProcessingTimeCallback callback) {
+
+			return timestamp -> {
+				if (status.compareAndSet(STATUS_PENDING, STATUS_EXECUTING)) {
+
+					try {
+						callback.onProcessingTime(timestamp);
+
+						status.compareAndSet(STATUS_EXECUTING, STATUS_SUCCEEDED);
+					} catch (Throwable throwable) {
+						// this may be a true execution failure or a cancellation
+						// that can be interrupted to running
+
+						status.compareAndSet(STATUS_EXECUTING, STATUS_FAILED);
+						throw throwable;
+					} finally {
+						onDoneActionFuture.complete(this);
+					}
+				}
+			};
+		}
+
+		private ProcessingTimeCallback wrapPeriodicCallback(ProcessingTimeCallback callback) {
+
+			return timestamp -> {
+				if (status.compareAndSet(STATUS_PENDING, STATUS_EXECUTING)) {
+
+					callback.onProcessingTime(timestamp);
+					if (!status.compareAndSet(STATUS_EXECUTING, STATUS_PENDING)) {
+						checkState(status.get() == STATUS_CANCELLED);
+						onDoneActionFuture.complete(this);
+					}
+				} else if (status.get() == STATUS_CANCELLED) {
+					// when the underlying future is not successfully cancelled, this ensures that it
+					// will not run at idle all the time
+					// (see the logic of canceling from the "STATUS_PENDING" state in #cancel())
+					checkState(wrappedFuture.isDone());
```

Comment:
race condition: `CAS status`  and `future.cancel()` is not a single (atomic) operation